DEV-297 - Updated integration tests to be more automatic given env variables

acanedo opened the pull request:
- Added `dotenv` package to be able to use .env file in root directory.
- Added TEST.md explaining how to set environment variables for testing.
- Modified integration tests to use environment variables to an specific algorithmia account and api address.
- Added three more methods to AlgorithmiaClient.ts to be able to build and publish algorithms and delete organization. This helps automate all tests without having to do preloading of data within an account.
- Added lint-fix npm script to also handle and clean the test files.
- Renamed some testing files and static content to have more meaning.
- Created a TestUtils.ts file to keep reusable code needed for all the test suites.

Commit:
DEV-297 - Updated integration tests to be more automatic given env variables

Commit:
code clean up and added TEST.md

Commit:
lint

warmbowski commented on test/TEST.md:
```diff
@@ -0,0 +1,34 @@
+   |ALGORITHMIA_TEST_DEFAULT_API_KEY|`<key for Algorithmia api url>`|Required for your algorithmia testing client|
+   |ALGORITHMIA_TEST_API_ADDRESS|`<secret key to run Algorithmia api>`|Required for your algorithmia testing client|
+   |ALGORITHMIA_TEST_USERNAME|`<your Algorithmia test account username>`|Required to test algorithmia client methods, file uploads, and directory creation|
+   |ALGORITHMIA_TEST_USERNAME|`<secret key to your admin Algorithmia account`|Required to test organization read, creation and update|
```
Probably meant to be `ALGORITHMIA_TEST_ADMIN_API_KEY`

acanedo replied:
yes!, good catch fixed.

warmbowski commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
 node_modules/
-
+.env
```
Maybe it would be helpful to create a file called `.env.example` and fill it with bogus data and a comment NOT to use the file but to copy it into a new `.env` file and change the values, and commit it to the repo. Just to make setting up a .env file a bit easier.

acanedo replied:
I made .env.example and its comment. Do we still want to gitignore `.env` I'd assume so right?

warmbowski replied:
yeah. I love that you set that up and git ignored it. That's a perfect solution so folks don't push their keys to the repo but still can set up easily.

warmbowski commented on test/TestUtils.ts:
```diff
@@ -0,0 +1,48 @@
+dotenv.config();
+
+const envMessage = (envVar) =>
+  `Environment variable ${envVar} is required to run this test suite. Reference TEST.md to help set up this variable`;
```
Nice!

warmbowski commented on package.json:
```diff
@@ -14,7 +14,7 @@
     "prepare": "npm run compile",
     "prepublishOnly": "npm test --detectOpenHandles",
-    "lint": "eslint ./src/**/*.{js,ts} --quiet",
+    "lint": "eslint ./src/**/*.{js,ts} ./test/**/*.{js,ts,test.ts,test.js} --quiet",
```
nice!

warmbowski commented on test/TEST.md:
```diff
@@ -0,0 +1,34 @@
+   |ALGORITHMIA_TEST_USERNAME|`<your Algorithmia test account username>`|Required to test algorithmia client methods, file uploads, and directory creation|
+   |ALGORITHMIA_TEST_USERNAME|`<secret key to your admin Algorithmia account`|Required to test organization read, creation and update|
+
+Normally you'd want `ALGORITHMIA_TEST_API_ADDRES` to be `https://test.algorithmia.com` or some sort of testing environment. 
```
small typo here. also, let's leave the test url out.
```suggestion
Normally you'd want `ALGORITHMIA_TEST_API_ADDRESS` to be some sort of testing environment. 
```

acanedo replied:
fixed

warmbowski reviewed (dismissed):
This is great! And worked like a charm. Just a change in TEST.md.
<img width="513" alt="Screen Shot 2021-06-03 at 8 05 57 AM" src="https://user-images.githubusercontent.com/1071557/120669746-9580d280-c444-11eb-95f0-1e38ddd929a9.png">

Commit:
addressed comments

acanedo commented:
Was going through this and realized I had a bug. Would you mind re-approving/re-reviewing.

warmbowski reviewed (commented):
Still works fine and the code LGTM. 

One thing I noticed is that the orgs that are made aren't cleaned up. 

Two things I just noticed:
* Not a big deal if it cannot do it, but is it possible to remove the org at the end of the testing?
* change the bogus emails to use `@example.com`, as that is guaranteed not to be a real email server (in case any verification email is sent)

<img width="1310" alt="Screen Shot 2021-06-03 at 12 51 17 PM" src="https://user-images.githubusercontent.com/1071557/120703902-c1af4a00-c46a-11eb-9544-e0e95576cffe.png">

Commit:
addressed comments, changed bogus emails to @example.com

warmbowski commented on test/AlgorithmiaClient.test.ts:
```diff
@@ -125,66 +189,104 @@ describe('Localisation initialization', () => {
+        ALGORITHMIA_TEST_API_ADDRESS
+      ).getOrganization(createdOrganization.org_name);
 
-      expect(organization.org_name).toBe('MyOrg1614039696593');
```
Oh yeah. I see. Well there is a DELETE endpoint on `/v1/organizations/:orgname`, but I have no idea if we can add that to this node lib. Also, the other odd thing is that if you delete an org, it just gets marked as deleted, but not removed from the database, so you still cannot recreate the org with the same name. Lets just go with this for now as you have it.  In the mean time, we can figure out later who to talk to about adding the org delete endpoint. Does that sound good?

warmbowski replied:
Maybe create a ticket about adding that DELETE endpoint so we can get some feedback on it.

acanedo commented:
@warmbowski So I did some investigating and I found

`DELETE         /v1/organizations/:orgName                                                              @controllers.v1.Organizations.deleteOrganization(orgName: String)`

but DIDN'T work :(.  Or it behaves strangely.

So I'm not sure if the endpoint actually just does a soft delete on an organization. 

But if I run a `delete by org_name` and then try and run a `getOrganization` then it tells me there's no org by that name. That seems correct. 

The problem is that if I try and create a new org with the `same org_name` (as the one previously deleted) then the response tells me the org_name is already taken. Which doesn't seem right given I deleted the organization already. I'm not sure if this intended by design? 

Anyways, I also fixed the bogus emails to have @example.com and left the organization tests to NOT delete instead.

warmbowski commented:
> The problem is that if I try and create a new org with the `same org_name` (as the one previously deleted) then the response tells me the org_name is already taken. Which doesn't seem right given I deleted the organization already. I'm not sure if this intended by design?

From what I have heard, this is by design.

Commit:
address comments added a way to delete the test orgs after a test

acanedo commented:
pushed new changes, it now deletes orgs.
